modelhub/modelhub/models/funnel_discovery.py

```diff
@@ -0,0 +1,414 @@
+            )
+            sort_nice_names_by = [data[idx] for idx in partition.index_columns]
+
+        # always sort by moment, since we need to respect the order of the nice names in the data
+        # for getting the correct navigation paths based on event time
+        sort_nice_names_by += [data['moment']]
+
+        nice_name = _location_stack.ls.nice_name.sort_by_series(by=sort_nice_names_by)
+        agg_steps = nice_name.to_json_array(partition=partition)
+        flattened_lc, offset_lc = agg_steps.json.flatten_array()
+
+        # flattening will assume items are still json type, we need to extract the items
+        # as scalar types (string items)
+        if is_bigquery(data.engine):
+            flattened_lc = flattened_lc.copy_override(
+                expression=bach.expression.Expression.construct('JSON_EXTRACT_SCALAR({})', flattened_lc)
+            )
+        elif is_postgres(data.engine):
+            flattened_lc = flattened_lc.copy_override(
+                expression=bach.expression.Expression.construct(
+                    '{} #>> {}',
+                    flattened_lc,
+                    bach.expression.Expression.raw("'{}'")
+                )
+            )
+
+        flattened_lc = flattened_lc.astype('string')
+
+        offset_lc = offset_lc.copy_override(name='__root_step_offset')
+
+        nav_df = flattened_lc.to_frame()
+        nav_df[offset_lc.name] = offset_lc
+
+        nav_df = nav_df.sort_values(
+            by=nav_df.index_columns + [offset_lc.name],
+            # since we are using lag, we should reverse the step order
+            ascending=[True] * len(nav_df.index_columns) + [False],
+        )
+
+        window = nav_df.groupby(by=nav_df.index_columns).window()
+        root_step_series = window[flattened_lc.name]
+
+        all_step_series = {}
+        for step in range(1, steps + 1):
+            step_series_name = f'{flattened_lc.name}_step_{step}'
+            if step == 1:
+                next_step = root_step_series.copy_override(group_by=None)
+            else:
+                next_step = root_step_series.window_lag(offset=step - 1)
+
+            all_step_series[step_series_name] = (
+                next_step.copy_override(name=step_series_name)
+            )
+
+        result = nav_df.copy_override(
+            base_node=window.base_node,
+            series={
+                **all_step_series,
+                offset_lc.name: offset_lc.copy_override(base_node=window.base_node),
+            }
+        )
+        result = result.materialize(node_name='step_extraction')
+
+        # limit rows
+        if n_examples is not None:
+            result = result[result[offset_lc.name] < n_examples]
+
+        # re-order rows
+        result = result.sort_values(by=result.index_columns + [offset_lc.name])
+        result = result.drop(columns=[offset_lc.name])
+        return result
+
+    @staticmethod
+    def tag_step(step_series: bach.Series) -> bach.Series:
```

Comment:
let's make this private
```suggestion
    def _tag_step(step_series: bach.Series) -> bach.Series:
```